src/StorageFlattening.cpp

```diff
@@ -259,7 +260,7 @@ class FlattenDimensions : public IRMutator {
             Expr store = Call::make(value.type(), Call::image_store,
                                     args, Call::Intrinsic);
             return Evaluate::make(store);
-        } else if (in_gpu && textures.count(op->name)) {
+        } else if (in_gpu && textures.count(op->name) && false && in_device_api != DeviceAPI::CUDA) { // CUDA writes are still directly to memory
```

Comment:
I'm not sure what to do here.  In CUDA the texture sampling is read-only, so writes should be flattened as before.  But when the StorageFlattening pass runs, DeviceAPI hasn't been assigned and so there isn't anything to key off to flatten or not flatten at this phase.

There should also be a check somewhere to ensure there are not reads and writes to a buffer-as-texture within any one kernel.

Reply:
Moved the DeviceAPI assignment to before StorageFlattening, I need it anyways so allocations can be properly strided if necessary.